openmls/src/group/managed_group/validation.rs

```diff
@@ -0,0 +1,244 @@
+            InboundMessage::Plaintext(plaintext) => {
+                // We expect a membership tag for plaintext messages
+                // TODO #106: Membership tag is not expected for external senders
+                membership_tag_required = true;
+                (plaintext, vec![])
+            }
+        };
+
+        // Check that the sender is a valid member of the tree
+        // The sender index must be within the tree and the corresponding leaf node must not be blank
+        let sender = plaintext.sender().to_leaf_index();
+        if sender > self.group.tree().leaf_count() || self.group.tree().nodes[sender].is_blank() {
+            return Err(ManagedGroupError::InvalidMessage(
+                InvalidMessageError::UnknownSender,
+            ));
+        }
+
+        // Extract the credential
+        // Unwrapping here is safe, because we know the leaf node exists and is not blank
+        let credential = self.group.tree().nodes[sender]
+            .key_package
+            .as_ref()
+            .unwrap()
+            .credential()
+            .clone();
+
+        // Get the serialized group context for further verification
+        let serialized_context = self.group.context().tls_serialize_detached().map_err(|_| {
+            ManagedGroupError::LibraryError("Could not serialize group context".into())
+        })?;
+
+        // Verifiy the membership tag if needed
+        if membership_tag_required {
+            if !plaintext.has_membership_tag() {
+                return Err(ManagedGroupError::InvalidMessage(
+                    InvalidMessageError::MissingMembershipTag,
+                ));
+            }
+            plaintext
+                .verify_membership(
+                    backend,
+                    serialized_context.as_slice(),
+                    self.group.epoch_secrets().membership_key(),
+                )
+                .map_err(|_| {
+                    ManagedGroupError::InvalidMessage(InvalidMessageError::MembershipTagMismatch)
+                })?;
+        }
+
+        // Check that if the message is a commit the confirmation tag is present
+        if plaintext.is_commit() && plaintext.confirmation_tag().is_none() {
+            return Err(ManagedGroupError::InvalidMessage(
+                InvalidMessageError::MissingConfirmationTag,
+            ));
+        }
+
+        Ok(UnverifiedMessage {
+            plaintext,
+            credential,
+            aad,
+        })
+    }
+
+    /// This processing function does most of the semantic verifications.
+    /// It returns a [ProcessedMessage] enum.
+    pub fn process_unverified_message(
+        &mut self,
+        message: UnverifiedMessage,
+        signature_key: Option<SignaturePublicKey>,
+        backend: &impl OpenMlsCryptoProvider,
+    ) -> Result<ProcessedMessage, ManagedGroupError> {
+        /*
+         - Signature verification, either with leaf key or optional parameter
```

Comment:
If I read this function correctly, some of these things are not actually checked (yet), are they?